File: scripts/python/src/tools/dotfile/secret/doctor.py
import os
import platform

from tools.core.console import colors_enabled
from tools.core.process import capture, silent
from tools.dotfile.report import emit, heading, plural, row
from tools.dotfile.secret.canaries import all_canaries, canaries_file
from tools.dotfile.secret.identity import (
    have,
    identity_path,
    mode_of,
    public_key,
    sops_env,
    stray_paths,
)
from tools.dotfile.secret.keys import (
    is_recovery,
    keys_file,
    label_for,
    load_recipients,
    recovery_labels,
    sops_drifted,
    sops_file,
)
from tools.dotfile.secret.scan import encrypted_paths
from tools.dotfile.state import log, shorten

HOOKS = ("pre-commit", "pre-push")


def tools_row():
    missing = [name for name in ("age", "age-keygen", "sops") if not have(name)]
    if missing:
        return row("bad", "tools", "not on PATH", [(name, "") for name in missing])
    return row("ok", "tools", "age and sops present")


def identity_row(ctx):
    path = identity_path(ctx)
    if not os.path.isfile(path):
        return row(
            "bad", "identity", f"missing: {shorten(ctx, path)}", [("dotfile secret init", "")]
        )
    mode = mode_of(path)
    if mode & 0o077:
        return row(
            "bad",
            "identity",
            f"readable beyond you ({mode:04o})",
            [(f"chmod 600 {shorten(ctx, path)}", "")],
        )
    if not public_key(path):
        return row("bad", "identity", f"unreadable as an age key: {shorten(ctx, path)}")
    return row("ok", "identity", f"{shorten(ctx, path)} (0600)")


def enrolled_row(ctx, recipients):
    key = public_key(identity_path(ctx))
    if not key:
        return row("note", "enrolled", "no identity to check")
    label = label_for(recipients, key)
    if label:
        return row("ok", "enrolled", f"this machine is '{label}'")
    return row(
        "bad",
        "enrolled",
        "this machine is not a recipient yet",
        [
            ("on a machine that already decrypts, then push:", ""),
            (f"dotfile secret enroll {suggested_label()} {key}", ""),
            ("or here, with the recovery key:", ""),
            (f"dotfile secret enroll {suggested_label()} --using <recovery>", ""),
        ],
    )


def suggested_label():
    name = platform.node().split(".")[0].lower()
    cleaned = "".join(char for char in name if char.isalnum() or char in "._-")
    return cleaned or "<label>"


def recipients_row(ctx, recipients):
    if not recipients:
        return row("bad", "recipients", f"none in {shorten(ctx, keys_file(ctx))}")
    items = [(label, "recovery" if is_recovery(label) else "") for label in sorted(recipients)]
    if not recovery_labels(recipients):
        return row(
            "bad",
            "recipients",
            f"{len(recipients)} enrolled, none named recovery*",
            items + [("one lost disk loses everything encrypted", "")],
        )
    return row("ok", "recipients", f"{len(recipients)} enrolled", items)


def sops_row(ctx, recipients):
    if not recipients:
        return row("note", ".sops.yaml", "nothing to generate yet")
    if sops_drifted(ctx, recipients):
        return row(
            "bad",
            ".sops.yaml",
            "does not match config/keys.dotfile",
            [("dotfile secret sync", "")],
        )
    return row("ok", ".sops.yaml", f"matches {os.path.basename(keys_file(ctx))}")


def canaries_row(ctx):
    path = canaries_file(ctx)
    local = os.path.isfile(path)
    if local:
        mode = mode_of(path)
        if mode & 0o077:
            return row(
                "bad",
                "canaries",
                f"readable beyond you ({mode:04o})",
                [(f"chmod 600 {shorten(ctx, path)}", "")],
            )
    active, _notes = all_canaries(ctx)
    if not active:
        return row(
            "warn",
            "canaries",
            "no private values to match against",
            [("dotfile secret edit vars.enc.yaml", "")],
        )
    items = []
    if local:
        items.append(
            (shorten(ctx, path), "local only; this machine, not the others"),
        )
    return row("ok", "canaries", f"{plural(len(active), 'value')} guarded", items)


def hooks_row(ctx):
    result = capture(["git", "-C", ctx.root, "config", "core.hooksPath"])
    configured = result.stdout.strip()
    if os.path.basename(configured) != ".githooks":
        return row("bad", "hooks", "core.hooksPath is not .githooks", [("./setup.sh", "")])
    broken = [
        name for name in HOOKS if not os.access(os.path.join(ctx.root, ".githooks", name), os.X_OK)
    ]
    if broken:
        return row("bad", "hooks", "not executable", [(name, "") for name in broken])
    return row("ok", "hooks", "pre-commit and pre-push active")


def decryptable(ctx, path):
    return silent(["sops", "-d", path], cwd=ctx.root, env=sops_env(ctx)) == 0


def sealed_row(ctx):
    paths = encrypted_paths(ctx)
    if not paths:
        return row("note", "sealed", "no encrypted files yet")
    counted = plural(len(paths), "file")
    if not have("sops"):
        return row("bad", "sealed", f"{counted}, sops missing")
    locked = [path for path in paths if not decryptable(ctx, path)]
    if locked:
        return row(
            "bad",
            "sealed",
            f"{len(locked)} of {counted} will not decrypt here",
            [(path, "") for path in locked],
        )
    return row("ok", "sealed", f"{counted}, all decrypt here")


def git_config(ctx, name):
    return capture(["git", "-C", ctx.root, "config", name]).stdout.strip()


def diffs_row(ctx):
    if not os.path.isfile(os.path.join(ctx.root, ".gitattributes")):
        return row("warn", "diffs", "no .gitattributes, encrypted files diff as ciphertext")
    if git_config(ctx, "diff.sops.cachetextconv") == "true":
        return row(
            "bad",
            "diffs",
            "cachetextconv would write plaintext into .git",
            [("git config diff.sops.cachetextconv false", "")],
        )
    if not git_config(ctx, "diff.sops.textconv"):
        return row("warn", "diffs", "sops textconv not configured", [("./setup.sh", "")])
    return row("ok", "diffs", "git diff decrypts sops files locally")


def stray_finding(recipients, mine, other):
    if not other:
        return "warn", "not readable as an age key"
    if other == mine:
        return "warn", "this machine's own key, duplicated here"
    label = label_for(recipients, other)
    if not label:
        return "note", "not a recipient here, so it opens nothing in this repository"
    if is_recovery(label):
        return "warn", f"the '{label}' key, which is meant to live off-machine"
    return "warn", f"the '{label}' key; that machine's identity, on the wrong machine"


def strays_row(ctx, recipients):
    found = [path for path in stray_paths(ctx) if os.path.isfile(path)]
    if not found:
        return row("ok", "strays", "no identity outside the state directory")
    mine = public_key(identity_path(ctx))
    findings = [(path, *stray_finding(recipients, mine, public_key(path))) for path in found]
    items = [(shorten(ctx, path), note) for path, _kind, note in findings]
    if any(kind == "warn" for _path, kind, _note in findings):
        return row("warn", "strays", "another age identity on this machine", items)
    return row("note", "strays", "another age identity, unrelated to this one", items)


def cmd_doctor(ctx, show_all):
    recipients = load_recipients(ctx)
    color_on = colors_enabled()
    heading("doctor", shorten(ctx, ctx.root), color_on)

    entries = [
        tools_row(),
        identity_row(ctx),
        enrolled_row(ctx, recipients),
        recipients_row(ctx, recipients),
        sops_row(ctx, recipients),
        sealed_row(ctx),
        canaries_row(ctx),
        hooks_row(ctx),
        diffs_row(ctx),
        strays_row(ctx, recipients),
    ]
    for entry in entries:
        emit(entry, color_on)

    bad = [entry for entry in entries if entry[0] == "bad"]
    if bad:
        log("")
        log(f"{len(bad)} of {len(entries)} checks failed")
        raise SystemExit(1)
    if show_all:
        log("")
        log(f"keys   {shorten(ctx, keys_file(ctx))}")
        log(f"sops   {shorten(ctx, sops_file(ctx))}")

File: scripts/rust/crates/dotfile/src/secret/doctor.rs
use super::{canaries, recipients, scan, sops, vault};
use crate::context::Context;
use std::fs;
use std::path::Path;
use std::process::ExitCode;

pub fn suggested_label(context: &Context) -> String {
    if let Some(host) = crate::hosts::this(context) {
        return host;
    }
    let host = context
        .env("HOSTNAME")
        .map(|v| v.to_string_lossy().into_owned())
        .or_else(|| {
            let mut command = context.command("hostname");
            sops::capture(&mut command, 4096, "hostname")
                .ok()
                .map(|s| String::from_utf8_lossy(&s).into_owned())
        })
        .unwrap_or_else(|| "machine".into());
    let label: String = host
        .split('.')
        .next()
        .unwrap_or("machine")
        .to_ascii_lowercase()
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || "._-".contains(*c))
        .collect();
    if recipients::valid_label(&label) {
        label
    } else {
        "machine".into()
    }
}

pub fn run(context: &Context, all: bool) -> Result<ExitCode, String> {
    let recipients = recipients::load(context)?;
    let path = vault::identity_path(context);
    let mine = sops::public_key(context, &path).ok();
    let mut bad = 0;
    let mut row = |kind: &str, label: &str, detail: String| {
        println!("  {kind:<5} {label:<12} {detail}");
        if kind == "bad" {
            bad += 1;
        }
    };
    let missing: Vec<_> = ["age", "age-keygen", "sops"]
        .into_iter()
        .filter(|p| context.program(p).is_none())
        .collect();
    row(
        if missing.is_empty() { "ok" } else { "bad" },
        "tools",
        if missing.is_empty() {
            "age and sops present".into()
        } else {
            format!("not on PATH: {}", missing.join(" "))
        },
    );
    let private = path.is_file() && vault::mode_of(&path).is_ok_and(|mode| mode & 0o077 == 0);
    row(
        if private && mine.is_some() {
            "ok"
        } else {
            "bad"
        },
        "identity",
        if private && mine.is_some() {
            format!("{} (0600)", path.display())
        } else {
            format!(
                "missing, unreadable, or permissive identity: {}",
                path.display()
            )
        },
    );
    let label = recipients
        .iter()
        .find(|(_, key)| Some(*key) == mine.as_ref())
        .map(|(label, _)| label);
    row(
        if label.is_some() { "ok" } else { "bad" },
        "enrolled",
        label
            .map(|label| format!("this machine is '{label}'"))
            .unwrap_or_else(|| {
                format!(
                    "not a recipient yet; on a machine that already decrypts, or here with the recovery key: dotfile secret enroll {} --using <recovery>",
                    suggested_label(context)
                )
            }),
    );
    if let Some(label) = label {
        let wrapped = super::wrap::path(context, label);
        row(
            if wrapped.is_file() { "ok" } else { "warn" },
            "wrapped",
            if wrapped.is_file() {
                wrapped
                    .strip_prefix(&context.root)
                    .unwrap_or(&wrapped)
                    .display()
                    .to_string()
            } else {
                "not wrapped; run dotfile secret wrap".into()
            },
        );
    }
    let recovery = recipients
        .keys()
        .any(|label| recipients::is_recovery(label));
    row(
        if recipients.is_empty() || !recovery {
            "bad"
        } else {
            "ok"
        },
        "recipients",
        format!(
            "{} enrolled{}",
            recipients.len(),
            if recovery {
                " including recovery"
            } else {
                ", none named recovery*"
            }
        ),
    );
    let policy_matches = fs::read_to_string(context.root.join(".sops.yaml")).unwrap_or_default()
        == recipients::policy(&recipients);
    row(
        if policy_matches { "ok" } else { "bad" },
        ".sops.yaml",
        if policy_matches {
            "matches config/keys.dotfile".into()
        } else {
            "does not match config/keys.dotfile; run dotfile secret sync".into()
        },
    );
    let paths = scan::encrypted_paths(context)?;
    let mut locked = Vec::new();
    for path in &paths {
        if sops::decrypt(context, &context.root.join(path), None, false).is_err() {
            locked.push(path.display().to_string());
        }
    }
    row(
        if locked.is_empty() { "ok" } else { "bad" },
        "sealed",
        format!(
            "{} encrypted, {} will not decrypt here{}",
            paths.len(),
            locked.len(),
            if all && !locked.is_empty() {
                format!(": {}", locked.join(" "))
            } else {
                String::new()
            }
        ),
    );
    let (canaries, notes) = canaries::load(context)?;
    row(
        if notes.iter().any(|n| n.contains("readable beyond")) {
            "bad"
        } else if canaries.is_empty() {
            "warn"
        } else {
            "ok"
        },
        "canaries",
        format!(
            "{} private values guarded{}",
            canaries.len(),
            if notes.is_empty() {
                String::new()
            } else {
                format!(": {}", notes.join("; "))
            }
        ),
    );
    let config = |name: &str| {
        scan::git(context, &["config", "--get", name])
            .ok()
            .map(|v| String::from_utf8_lossy(&v).trim().to_string())
            .unwrap_or_default()
    };
    let hookspath = config("core.hooksPath");
    let executable = |path: &Path| {
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            path.metadata()
                .is_ok_and(|m| m.is_file() && m.permissions().mode() & 0o111 != 0)
        }
        #[cfg(not(unix))]
        {
            path.is_file()
        }
    };
    let hooks = Path::new(&hookspath)
        .file_name()
        .is_some_and(|n| n == ".githooks")
        && ["pre-commit", "pre-push"]
            .iter()
            .all(|hook| executable(&context.root.join(".githooks").join(hook)));
    row(
        if hooks { "ok" } else { "bad" },
        "hooks",
        if hooks {
            "pre-commit and pre-push active".into()
        } else {
            "core.hooksPath or executable hooks missing; run ./setup.sh".into()
        },
    );
    let cache = config("diff.sops.cachetextconv") == "true";
    let diffs = !config("diff.sops.textconv").is_empty();
    row(
        if cache {
            "bad"
        } else if diffs {
            "ok"
        } else {
            "warn"
        },
        "diffs",
        if cache {
            "cachetextconv would write plaintext into .git".into()
        } else if diffs {
            "local SOPS textconv configured".into()
        } else {
            "sops textconv not configured".into()
        },
    );
    let strays = [
        context
            .root_config
            .parent()
            .unwrap_or(&context.home)
            .join("sops/age/keys.txt"),
        context
            .home
            .join("Library/Application Support/sops/age/keys.txt"),
    ];
    let found: Vec<_> = strays
        .iter()
        .filter(|p| p.is_file())
        .map(|path| {
            let other = sops::public_key(context, path).ok();
            let (kind, note) = stray_finding(&recipients, mine.as_deref(), other.as_deref());
            (kind, format!("{}: {note}", path.display()))
        })
        .collect();
    row(
        if found.is_empty() {
            "ok"
        } else if found.iter().any(|(kind, _)| *kind == "warn") {
            "warn"
        } else {
            "note"
        },
        "strays",
        if found.is_empty() {
            "no identity outside the state directory".into()
        } else {
            found
                .iter()
                .map(|(_, note)| note.as_str())
                .collect::<Vec<_>>()
                .join("; ")
        },
    );
    if all {
        println!(
            "keys   {}\nsops   {}",
            context.root_config.join("keys.dotfile").display(),
            context.root.join(".sops.yaml").display()
        );
    }
    if bad != 0 {
        println!("{bad} of 10 checks failed");
    }
    Ok(if bad == 0 {
        ExitCode::SUCCESS
    } else {
        ExitCode::FAILURE
    })
}

pub(crate) fn stray_finding(
    recipients: &recipients::Recipients,
    mine: Option<&str>,
    other: Option<&str>,
) -> (&'static str, String) {
    let Some(other) = other else {
        return ("warn", "not readable as an age key".into());
    };
    if mine == Some(other) {
        return ("warn", "this machine's own key, duplicated here".into());
    }
    let Some((label, _)) = recipients.iter().find(|(_, key)| key.as_str() == other) else {
        return (
            "note",
            "not a recipient here, so it opens nothing in this repository".into(),
        );
    };
    if recipients::is_recovery(label) {
        (
            "warn",
            format!("the '{label}' key, which is meant to live off-machine"),
        )
    } else {
        (
            "warn",
            format!("the '{label}' key; that machine's identity, on the wrong machine"),
        )
    }
}
